### PR TITLE
Navigation page

### PR DESCRIPTION
**Pull request by hannesj**

### CODE REVIEW

File: app/store/location-store.coffee [hunk 1]
```diff
@@ -73,14 +79,26 @@ class LocationStore extends Store
   getLocationString: () ->
     "#{@address}::#{@lat},#{@lon}"
 
+  storeWatchId: (watchId) ->
+    @watchId = watchId
+
+  clearWatchId: ->
+    @watchId = undefined
+
+  getWatchId: ->
```
Comment (tpievila) [inline app/store/location-store.coffee]: Is the id meant to be stored in the caller (function doesn't bind this)?

File: app/util/geo-utils.coffee [hunk 2]
```diff
@@ -0,0 +1,11 @@
...
+toDeg = (rad) -> rad * 180 / Math.PI
+
+getBearing = (lat1,lng1,lat2,lng2) ->
+  lonScale = Math.cos toRad (lat1+lat2)/2
+  dy = lat2-lat1
+  dx = (lng2-lng1)*lonScale
+  (toDeg(Math.atan2(dx, dy)) + 360) % 360
+
+module.exports =
+  getBearing: getBearing
```
Comment (tpievila) [inline app/util/geo-utils.coffee]: lint:

  ⚡ app/util/geo-utils.coffee
     ⚡ #4: Spaces are required after commas. ,.
     ⚡ #4: Spaces are required after commas. ,.
     ⚡ #4: Spaces are required after commas. ,.
     ⚡ #5: Operators must be spaced properly. +.
     ⚡ #5: Operators must be spaced properly. /.
     ⚡ #6: Operators must be spaced properly. -.
     ⚡ #7: Operators must be spaced properly. -.
     ⚡ #7: Operators must be spaced properly. *.

File: app/component/map/navigation-map.cjsx [hunk 3]
```diff
@@ -0,0 +1,124 @@
...
+  componentWillUnmount: ->
+    @context.executeAction LocateActions.stopLocationWatch
+    if @intervalId
+      clearInterval(@intervalId)
+      @intervalId = undefined
+
+  render: ->
+    <div id="map" className="fullscreen"/>
+
+module.exports = NavigationMap
```
Comment (tpievila) [inline app/component/map/navigation-map.cjsx]: lint

  ✗ app/component/map/navigation-map.cjsx
     ⚡ #65: Invoking a constructor without parens and with arguments.
     ⚡ #77: Invoking a constructor without parens and with arguments.
     ✗ #87: Line exceeds maximum allowed length. Length is 105, max is 100.
     ⚡ #89: Invoking a constructor without parens and with arguments.

File: app/component/map/navigation-map.cjsx [hunk 4]
```diff
@@ -0,0 +1,124 @@
...
+class NavigationMap extends React.Component
+  @contextTypes:
+    getStore: React.PropTypes.func.isRequired
+    executeAction: React.PropTypes.func.isRequired
+
+  componentDidMount: ->
+    @context.executeAction LocateActions.startLocationWatch
+
+    require.ensure ['mapbox-gl/dist/mapbox-gl-dev', './mapbox-gl.css'], =>
+      require './mapbox-gl.css'
```
Comment (tpievila) [inline app/component/map/navigation-map.cjsx]: Shouldn't the callback take require as an argument?

Unusual import (not at the top, like everything else) is worthy of a comment too.

File: app/component/map/navigation-map.cjsx [hunk 5]
```diff
@@ -0,0 +1,124 @@
...
+          plan.legs[0].to.lat,
+          plan.legs[0].to.lon)
+
+      mapLoaded = new Promise (resolve) -> map.on 'load', resolve
+
+      Fulltilt = require('exports?window.FULLTILT!fulltilt/dist/fulltilt.js')
+      fulltiltLoaded = Fulltilt.getDeviceOrientation(type: 'world')
+
+      mapLoaded.then =>
+        #debug("map loaded")
```
Comment (tpievila) [inline app/component/map/navigation-map.cjsx]: No dead code into repos

File: app/component/map/navigation-map.cjsx [hunk 6]
```diff
@@ -0,0 +1,124 @@
...
+
+        @locationJSONsource = new mapboxgl.GeoJSONSource data: locationAsGeoJSON coordinates
+        map.addSource 'location', @locationJSONsource
+        map.addLayer getLayerforLocation()
+
+        fulltiltLoaded.then(
+          (orientation) => @initializeCompass(map, true, orientation)
+        , (errror) => @initializeCompass(map, false)
+        )
+    , 'mapboxgl'
```
Comment (tpievila) [inline app/component/map/navigation-map.cjsx]: Somewhat a PITA to see to which call this is a parameter for, when the actual call might not even fit on the display at the same time. A comment would be good.

Why do we need a named chunk here?

File: app/component/map/navigation-map.cjsx [hunk 7]
```diff
@@ -0,0 +1,124 @@
...
+      newCoordinates = @context.getStore('LocationStore').getLocationState()
+      if newCoordinates.lon != coordinates.lat or newCoordinates.lon != coordinates.lon
+        @locationJSONsource.setData locationAsGeoJSON coordinates
+        coordinates = newCoordinates
+      if compassAvailable
+        euler = orientation.getScreenAdjustedEuler()
+      map.easeTo
+        center: [coordinates.lat, coordinates.lon]
+        bearing: if compassAvailable then 360 - euler.alpha else coordinates.heading
+        pitch: if compassAvailable and isNaN(euler.beta) then Math.max(euler.beta, 0) else 45
```
Comment (tpievila) [inline app/component/map/navigation-map.cjsx]: Wait, what? If beta is not a number, take a max of it and 0?

File: app/page/navigation.cjsx [hunk 8]
```diff
@@ -0,0 +1,11 @@
+React              = require 'react'
+DefaultNavigation  = require '../component/navigation/default-navigation'
+NavigationMap      = require '../component/map/navigation-map'
+
+class NavigationPage extends React.Component
+  render: ->
+    <DefaultNavigation className="fullscreen">
+      <NavigationMap hash={parseInt(@props.params.hash)}/>
```
Comment (tpievila) [inline app/page/navigation.cjsx]: It's not obvious what the hash represents and what it's used for.

File: app/store/location-store.coffee [hunk 9]
```diff
@@ -46,6 +46,12 @@ class LocationStore extends Store
     @status = @STATUS_FOUND_LOCATION
     @emitChange()
 
+  storeLocationWithoutEmit: (location) ->
+    @lat = location.lat
+    @lon = location.lon
+    @heading = if location.heading then location.heading else @heading
+    @status = @STATUS_FOUND_LOCATION
+
```
Comment (tpievila) [inline app/store/location-store.coffee]: Would be good to comment here shortly why we sometimes want to emit and sometimes not.

This is pretty much the same as storeLocation, but heading is missing in the old function. If intentional, would be good to insert a comment on that.

Also, could the storeLocation simply call storeLocationWithoutEmit and then emit, to remove code duplication?

File: app/util/geo-utils.coffee [hunk 10]
```diff
@@ -0,0 +1,11 @@
+toRad = (deg) -> deg * Math.PI / 180
+toDeg = (rad) -> rad * 180 / Math.PI
+
+getBearing = (lat1,lng1,lat2,lng2) ->
```
Comment (tpievila) [inline app/util/geo-utils.coffee]: I think something short like:

    // Get direction of movement from two first to second point.
    // North is 0, east is 90, south 180, west 270

File: server/server.js [hunk 11]
```diff
@@ -32,6 +32,10 @@ function setUpStaticFolders() {
   app.use(rootPath +"css", express.static(cssFolder))
   var jsFolder = path.join(staticFolder, 'js')
   app.use(rootPath +"js", express.static(jsFolder))
+  var mapFontsFolder = path.join(staticFolder, 'map', 'fonts')
+  app.use(rootPath +"mapFonts", express.static(mapFontsFolder, {setHeaders: function(res){res.setHeader("Content-Encoding","gzip")}}))
+  var mapFolder = path.join(staticFolder, 'map')
+  app.use(rootPath +"map", express.static(mapFolder))
```
Comment (tpievila) [inline server/server.js]: We don't use jslint, but I think that one line is pretty long and that there should be whitespace around operators.

File: app/store/location-store.coffee (hunk 1 above)
Comment (hannesj) [inline app/store/location-store.coffee]: Fluxible binds it automatically

File: app/util/geo-utils.coffee (hunk 2 above)
Comment (hannesj) [inline app/util/geo-utils.coffee]: Fixed

File: app/component/map/navigation-map.cjsx (hunk 3 above)
Comment (hannesj) [inline app/component/map/navigation-map.cjsx]: Fixed

File: app/component/map/navigation-map.cjsx (hunk 4 above)
Comment (hannesj) [inline app/component/map/navigation-map.cjsx]: I changed the import logic quite a lot, now all of the requires are in the componentWillMount, where they are asynchronously required. The componentDidMount will then continue loading them after they have been downloaded.

File: app/component/map/navigation-map.cjsx (hunk 5 above)
Comment (hannesj) [inline app/component/map/navigation-map.cjsx]: Removed

File: app/component/map/navigation-map.cjsx (hunk 6 above)
Comment (hannesj) [inline app/component/map/navigation-map.cjsx]: The changed require process removed this line. Named chunks are nice with debugging.

File: app/component/map/navigation-map.cjsx (hunk 7 above)
Comment (hannesj) [inline app/component/map/navigation-map.cjsx]: Whoops, there seems to be a exclamation mark missing here after combining the two handlers into one. Fixed

File: app/page/navigation.cjsx (hunk 8 above)
Comment (hannesj) [inline app/page/navigation.cjsx]: Comment added

File: app/store/location-store.coffee (hunk 9 above)
Comment (hannesj) [inline app/store/location-store.coffee]: Fixed according to your suggestion, nicer that way.

File: app/util/geo-utils.coffee (hunk 10 above)
Comment (hannesj) [inline app/util/geo-utils.coffee]: Added

File: server/server.js (hunk 11 above)
Comment (hannesj) [inline server/server.js]: Fixed all the static file definitions.

**Pull request merged**